src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,173 @@
+                    throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
+
+                _thread = Thread.CurrentThread;
+
+                try
+                {
+                    try
+                    {
+                        // As soon as the Started flag is set, this thread may be aborted asynchronously
+                        if (Interlocked.CompareExchange(ref _state, (int)State.Started, (int)State.None) == (int)State.None)
+                        {
+                            t_execution = this;
+                            _action();
+                        }
+                    }
+                    finally
+                    {
+                        if ((_state & (int)State.Started) != 0)
+                        {
+                            // Set the Finished flag to prevent a potential subsequent AbortThread call
+                            State oldState = (State)Interlocked.Or(ref _state, (int)State.Finished);
+
+                            if ((oldState & State.AbortRequested) != 0)
+                            {
+                                // Either in SFAR or SFA state
+                                while (true)
+                                {
+                                    // The enclosing finally may be cloned by the JIT for the non-exceptional code flow.
+                                    // In that case this code is not guarded against a thread abort, so make this FCall as
+                                    // soon as possible.
+                                    bool resetAbortRequest = ResetAbortThread();
+
+                                    // If there is an Abort in progress, we need to wait until it sets the TS_AbortRequested
+                                    // flag on this thread, then we can reset the flag and safely exit this frame.
+                                    if (((oldState & State.RunningAbort) == 0) || resetAbortRequest)
+                                    {
+                                        break;
+                                    }
+
+                                    // It should take very short time for AbortThread to set the TS_AbortRequested flag
+                                    Thread.Sleep(0);
+                                    oldState = (State)Volatile.Read(ref _state);
+                                }
+                            }
+
+                            t_execution = null;
+                        }
+                    }
+                }
+                catch (ThreadAbortException) when (_cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException(_cancellationToken);
+                }
+            }
+
+            public void Abort()
+            {
+                // Prevent potential refetching of _state from shared memory
+                State curState = (State)Volatile.Read(ref _state);
```

Comment:
I wonder if most of this synchronization is really necessary?  CancellationToken.{Unsafe}Register returns a CancellationTokenRegistration, which guarantees that upon a call to its Dispose returning, either the callback will never run (because it was successfully unregistered prior to the callback being invoked) or the callback will have finished running.  That's done so that someone can do:
```C#
using (cancellationToken.Register(callback))
{
    StuffToBeCanceled();
}
```
and know that once that using block is exited, the callback won't be running and won't be invoked subsequently.

Could most/all of this synchronization go away and just rely on that? e.g. something like:
```C#
public void Run()
{
    if (t_execution)
    {
        throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
    }

    _thread = Thread.CurrentThread;
    try
    {
        t_execution = true;
        using (_cancellationToken.UnsafeRegister(e =>((Execution)e!).Cancel(), this))
        {
            _action();
        }
    }
    catch (ThreadAbortException tae)
    {
        Exception e = new OperationCanceledException(_cancellationToken);
        if (tae.StackTrace is string stackTrace)
        {
            ExceptionDispatchInfo.SetRemoteStackTrace(e, stackTrace);
        }
        throw e;
    }
    finally
    {
        t_execution = false;
        SpinWait sw = default;
        while (!ResetAbortThread())
        {
            sw.SpinOnce();
        }
    }
}

private void Cancel() => AbortThread(_thread.GetNativeHandle());
```
?

Reply:
Thank you for the thorough code review!  The original API allowed multiple simultaneous `Run` and `Abort` calls and I just moved its implementation to a private class.  It definitely may be simplified for the new API shape.  I will try your suggestions.

Reply:
Cool.  Yeah, I figured this was a holdover from when the API shape made things more complicated.

Reply:
> CancellationToken.{Unsafe}Register returns a CancellationTokenRegistration, which guarantees that upon a call to its Dispose returning, either the callback will never run (because it was successfully unregistered prior to the callback being invoked) or the callback will have finished running.

That is nice property; however, I am not sure it fits our purposes. A safe thread abort does not abort execution of `finally` and `catch` blocks, so `AbortThread` waits until the victim thread is outside of any such blocks.  That means we cannot have the `CancellationTokenRegistration.Dispose` call inside any `finally`/`catch` block, because that will lead to a deadlock.  And if we place this call naked inside the `try` block, it will not be executed in case the `_action()` call throws an exception.

The original approach avoids this deadlock by resetting the `TS_AbortRequested` flag, which indicates a successful abortion and causes the `AbortThread` call to return.

Reply:
> The original approach avoids this deadlock by resetting the TS_AbortRequested flag, which indicates a successful abortion and causes the AbortThread call to return.

The abort thread polls that? This can't be done with an extra try/finally around the callback inside the using that just sets that flag in a finally? (It can then be set again if necessary after the token is disposed).

Reply:
> The abort thread polls that?

Yes, it polls the flag in its retrying loop.

> This can't be done with an extra try/finally around the callback inside the using that just sets that flag in a finally?

Resetting the flag is already done in a `finally` block.  The only problem is that if the `Abort` has started executing, we may need to wait until it sets the flag.  I do not see how unregistering the callback would make that simpler.